feat: ark evaluator with langfuse

skokaina opened the pull request:
Major Features:
  - Langfuse + RAGAS integration in ark-evaluator
  - Introduce ark-evaluator as a single holistic evaluation service
  - update evaluator to handle context in langfuse adatper
  - refactor tests 

  Bug Fixes:
  - CI/CD reliability improvements
  - uv loop threading issue with next_asyncio for Ragas
  - Tool caching and navigation fixes
  - Documentation link fixes
  - Helm chart deployment issues

  Infrastructure:
  - Enhanced documentation and UI improvements

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [x] Update `./docs`
- [x] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [x] Linked issues #eg101

skokaina committed:
feat: ark-evaluator, integrate langfuse+ragas, merge evaluator-llm and evaluator-metric into one service

skokaina committed:
feat: ark-evaluator, integrate langfuse+ragas, merge evaluator-llm and evaluator-metric into one service

skokaina committed:
feat: ark-evaluator, integrate langfuse+ragas, merge evaluator-llm and evaluator-metric into one service

skokaina committed:
feat: ark-evaluator, integrate langfuse+ragas, merge evaluator-llm and evaluator-metric into one service

skokaina committed:
feat: ark-evaluator, integrate langfuse+ragas, merge evaluator-llm and evaluator-metric into one service

skokaina committed:
fix: unit tests for ark evaluator

skokaina committed:
chore: add missing tests

skokaina committed:
docs: update ark documentation

skokaina committed:
chore: remove evaluator-metric, merged now in ark-evaluator

skokaina committed:
chore: fix linting and docs

skokaina committed:
fix: dashboard polling and reference for queries, evaluator edits

skokaina committed:
feat: improve metric eval based view display

skokaina committed:
cicd: fix pipeline issues

iamhassaan committed:
fix:removed make langfuse-credentials and make langfuse-dashboard from docs

iamhassaan committed:
fix:re-added working make langfuse-credentials and make langfuse-dashboard

Commit:
fix: fix tests that broke as a result of merging the evaluator services

skokaina committed:
chore: fix display for evals

iamhassaan committed:
fix: make command ark-evaluator-dev

Commit:
fix: update the make test command to not fail if critical dependency issues are found

Commit:
fix: added helpful comment on builk file for npm audit

iamhassaan committed:
docs: enhanced local development guide for ark-evaluator

dwmkerr commented on ark/internal/controller/evaluation_controller.go:
```diff
@@ -349,13 +349,14 @@ func (r *EvaluationReconciler) resolveFinalParameters(ctx context.Context, evalu
-			Value: param.Value, // For now, we only support direct values from evaluators
+			Name:      param.Name,
+			Value:     param.Value,
+			ValueFrom: param.ValueFrom, // Include ValueFrom for proper resolution
```
Excellent!

dwmkerr commented on ark/internal/controller/evaluation_controller.go:
```diff
@@ -1064,6 +1073,38 @@ func (r *EvaluationReconciler) processEventEvaluation(ctx context.Context, evalu
 	return ctrl.Result{}, nil
 }
 
+// All complex context extraction logic moved to genai/context_retrieval_helper.go
```
Do need for breadcrumbs like this

iamhassaan replied:
addressed

dwmkerr commented on ark/internal/controller/evaluation_controller.go:
```diff
@@ -1064,6 +1073,38 @@ func (r *EvaluationReconciler) processEventEvaluation(ctx context.Context, evalu
+// All complex context extraction logic moved to genai/context_retrieval_helper.go
+
+// addContextToParameters adds contextual background information to evaluation parameters using the helper
+func (r *EvaluationReconciler) addContextToParameters(ctx context.Context, evaluation *arkv1alpha1.Evaluation, parameters map[string]string) map[string]string {
```
This looks useful is it worth adding to the docs about evaluation context with a brief snippet showing what you get?

iamhassaan replied:
added docs in : docs/content/reference/evaluations/evaluations.mdx

dwmkerr commented on ark/internal/genai/context_retrieval_helper.go:
```diff
@@ -0,0 +1,277 @@
+}
+
+// addMemoryContextIfAvailable extracts memory context into contextInfo
+func (h *ContextHelper) addMemoryContextIfAvailable(ctx context.Context, query *arkv1alpha1.Query, info *contextInfo) {
```
can you help explain what this does? also might be needed in the docs if we hydate with memory

iamhassaan replied:
@dwmkerr This function checks if a query has a memory reference and extracts
   context information about it for evaluation purposes.
   
   It doesn't actually hydrate (fetch) the conversation history from the
  memory service
  
  It only adds a reference noting that memory exists: "Previous conversation
   history available (stored at: http://postgres-memory:8080)"
   
   I've added documentation about this in the "Automatic Context Extraction"
  section of docs/content/reference/evaluations/evaluations.mdx
  
  @skokaina please validate this

skokaina replied:
yes I do, we don't fetch the actual data, that would kill the controller

dwmkerr commented on ark/internal/genai/context_retrieval_helper.go:
```diff
@@ -0,0 +1,277 @@
+
+	// Memory CRD only tracks address, actual conversation content is in external service
+	// For now, we note that conversation history exists at this address
+	// TODO: In future, could fetch actual conversation content from memory service
```
let's avoid todos maybe link to a jira ticket instead

iamhassaan replied:
added Jira: https://mckinsey.atlassian.net/browse/ARKQB-239

dwmkerr commented on samples/agent-modernization/agents/agent-java8-to-js-modernizer.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: ark.mckinsey.com/v1alpha1
```
I'm wondering if this should in fact go into the marketplace, we could then later add a helm chart for it, this'll keep the core repo cleaner? But open to ideas - this currently adds a lot of code which of course will have to be updated

iamhassaan replied:
@skokaina please advice here

skokaina replied:
@dwmkerr it's more a sample showing evaluations with positive and negative scenariis, the agent in itself is quite basic for the marketplace

dwmkerr commented on services/ark-dashboard/ark-dashboard/app/(dashboard)/evaluators/[name]/edit/page.tsx:
```diff
@@ -0,0 +1,160 @@
+"use client"
```
Can you add a video to the PR to show this in action?

dwmkerr commented on services/ark-dashboard/ark-dashboard/app/(dashboard)/query/[id]/page.tsx:
```diff
@@ -415,23 +428,29 @@ function QueryDetailContent() {
+  }, [loadQuery, isNew, targetTool, namespace])
+
+  // Auto-refresh for running query
+  useEffect(() => {
```
I have a big ask here, can we avoid auto refresh, and instead think about this:

- https://kubernetes.io/docs/reference/using-api/api-concepts/

This would mean we have in our ark-api:

`http://apiaddress/v1/namespace/default/query/name?watch=true`

The UI can then simply update with the fragments that come back. I'd *love* to setup this pattern properly but until we have it avoid polling as it is very ineffecient and I want to make 100% sure we move away from this pattern.

I would prefer that to be in a new PR but take polling out of this one as it'll be tech debt.

Please do feel free to slack chat what the spec of watch would be - its a standard k8s idiom that'd be a HUGE help, we can PR that capability incrementally, RBAC for `WATCH` and then the api layer then the UI

iamhassaan replied:
addressed ! Great comment!

skokaina replied:
suggest we do it on a separate ticker, @iamhassaan could you please follow-up with @Nab-0 to open one and reference it here ?

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/evaluation/enhanced-evaluation-detail-view.tsx:
```diff
@@ -108,25 +108,44 @@ export function EnhancedEvaluationDetailView({ evaluationId, namespace }: Enhanc
+  }, [loadEvaluation])
+
+  // Auto-refresh for running evaluation
+  useEffect(() => {
```
same here on watch

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/evaluation/evaluation-detail-view.tsx:
```diff
@@ -227,11 +256,45 @@ export function EvaluationDetailView({ evaluationId, namespace, enhanced = false
+  )
+  
+  // Check if this is a quality-based evaluation (LLM assessments)
+  const isQualityEvaluation = !isEventEvaluation && !isMetricsEvaluation && hasMetadata && (
```
a lot of this should be in evaluator constants file please

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/evaluation/evaluation-status-indicator.tsx:
```diff
@@ -77,21 +77,43 @@ export function EvaluationStatusIndicator({
+      const data = await evaluationsService.getEvaluationSummary(namespace, queryName, enhanced)
+      setSummary(data)
+    } catch (error) {
+      // If API fails, set default "no evaluations" state
```
should this be more like an error?

iamhassaan replied:
added
It shows an error now.
<img width="1549" height="626" alt="image" src="https://github.com/user-attachments/assets/eaf08b6a-6aa5-4c6a-9886-eebc84e1b823" />

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/evaluation/evaluation-status-indicator.tsx:
```diff
@@ -77,21 +77,43 @@ export function EvaluationStatusIndicator({
+  }, [loadSummary])
+
+  // Auto-refresh when evaluations are pending
+  useEffect(() => {
```
same on watch

iamhassaan replied:
addressed

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/sections/evaluations-section.tsx:
```diff
@@ -99,33 +99,52 @@ export const EvaluationsSection = forwardRef<{ openAddEditor: () => void }, Eval
+    }, [loadEvaluations])
+
+    // Auto-refresh for running evaluations
+    useEffect(() => {
```
same

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/sections/evaluators-section.tsx:
```diff
@@ -50,31 +50,22 @@ export const EvaluatorsSection = forwardRef<{ openAddEditor: () => void }, Evalu
-          description: `Successfully created ${createRequest.name}`
-        })
-      }
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
why not `_` and make sure eslint ignores those params, that's more idiomatic

iamhassaan replied:
done

iamhassaan replied:
addressed!

dwmkerr commented on services/ark-dashboard/ark-dashboard/components/sections/queries-section.tsx:
```diff
@@ -39,29 +39,48 @@ export const QueriesSection = forwardRef<{ openAddEditor: () => void }, QueriesS
+  }, [loadQueries]);
+
+  // Auto-refresh for running queries
+  useEffect(() => {
```
same

dwmkerr commented on services/ark-dashboard/ark-dashboard/dev.log:
```diff
@@ -0,0 +1,150 @@
+
```
please remove this

dwmkerr commented on services/ark-dashboard/ark-dashboard/lib/services/evaluations.ts:
```diff
@@ -316,23 +316,46 @@ export const evaluationsService = {
-        const nameWithoutSuffix = evaluationName.replace(/-eval$/, '')
-        return nameWithoutSuffix.includes(queryNameLower)
-      })
+      // We need to fetch detailed info for each evaluation to check the queryRef
```
this seems like a lot of chatter should it not be in `ark-api`?

iamhassaan replied:
Excellent recom.

I've implemented a proper backend filter to address this:

1. **Backend (ark-api)**: Added `query_ref` parameter to
  `/v1/namespaces/{namespace}/evaluations` endpoint that filters evaluations
  server-side by `spec.config.queryRef.name`
  
  2. **Frontend (ark-dashboard)**: Simplified `getByQueryRef()` to use the new
   filter parameter, eliminating the N+1 pattern

Commit:
feat: updated the polling in queries page to use react-query

Commit:
feat: remove polling for refresh buttons

skokaina committed:
Merge remote-tracking branch 'origin/main' into feat/ark-evaluator-with-langfuse

iamhassaan committed:
feat: improve evaluation system efficiency and error handling

  - Remove polling in favor of future Kubernetes watch pattern
  - Move evaluation filtering from frontend to backend API
  - Add query_ref parameter to evaluations endpoint for server-side filtering
  - Show error states when evaluation API calls fail
  - Fix eslint unused variables with underscore convention
  - Remove accidental dev.log file

Nishanth-McK commented on services/ark-api/ark-api/src/ark_api/api/v1/evaluations.py:
```diff
@@ -38,12 +38,23 @@
+            for item in result:
+                item_dict = item.to_dict()
+                # Check if this evaluation has a queryRef that matches
+                if (item_dict.get('spec', {}).get('config', {}).get('queryRef', {}).get('name') == query_ref):
```
I do not think we should do like a hard match against the `query_ref`. Do the `in` check. As a user maybe we will filter like all the token queries evaluations for example.

iamhassaan replied:
Good point @Nishanth-McK  about flexible filtering in general! However, for these specific
   functions (getByQueryRef and getEvaluationSummary), we actually need exact
  matching since we're getting evaluations for one specific query name.

  For example, when showing the evaluation status indicator for query
  "token-validation", we only want evaluations for that exact query - not for
  "token-validation-v2" or "my-token-validation".

  The partial matching approach would be great for a search/filter feature
  where users browse evaluations, but for these query-specific functions,
  exact matching ensures we show the right status. What do you think?

skokaina committed:
chore: merge

dwmkerr commented on services/ark-dashboard/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM node:24-alpine AS base
+FROM node:22-alpine AS base
```
why not 24?

skokaina replied:
the build was failing with node:24 - will retry

skokaina committed:
chore: update node version for ark-dashboard

skokaina committed:
chore: merge main

skokaina committed:
chore: fix merge and build
